pip/pip-269.md

```diff
@@ -0,0 +1,242 @@
+- **Cursor Epoch**: a counter to record how many times the the **Cursor Position Reset** executed. If it is increased after a in-flight reading, discard this reading.
+  - This attribute can help to prevent such error: a reading is started(`${Cursor position is resetting}` is `false`); a **Cursor position is resetting** is executed; the reading is completed(`${Cursor position is resetting}` is `false`).
+
+### Guarantee safe changes
+- Inside Cursor: the changes of attributes `read position`, `mark deleted position`, `cursor epoch`, and `cursor position is resetting` use the same lock.
+- Outside Cursor
+  - The action **Cursor Position Reset*** update `${Cursor position is resetting}` first, which guarantees that no other **Cursor Position Reset*** events will reenter(fast fail), and new reading can be rejected (In other words, it is the same as a non-reentrant lock).
+  - Each component (such as Dispatcher, Replicator...) modifies its attributes(each component to decide whether it needs to a locked).
+  - End the change: revert `${Cursor position is resetting}`; update `read position` and `mark deleted position`; increase **Cursor Epoch**(help to prevent such error: a reading is started(`${Cursor Epoch}` is increased); a **Cursor position is resetting** is executed; the reading is completed).
+  
+# Detailed Design
+
+**ManagedLedgerException.java**
+```java
+/** If a reading was injected by the event Cursor Position Reset, it will get a CursorModifyingException. **/
+public static class CursorModifyingException extends ManagedLedgerException {};
+```
+
+**CursorEpoch.java**
+```java
+/** A counter to record how many times the the **Cursor Position Reset** executed. If it is increased after a in-flight reading, discard this reading. **/
+private volatile int epoch;
+/**
+ * Indicate if there is a **Cursor Position Reset** running. Inject new readingg and discard in-flight reading if it is `true`.
+ * 0: no task is running.
+ * 1: the task is running 
+ */
+private volatile int isResetting;
+
+/**
+* Set `isResetting` to `true`.
+* return `-1` if there a running epoch increase task is in progress, else return {@link #epoch}.
+*/
+public synchronized int startReset();
+
+/**
+ * Compare and set `isResetting` to `false`; increase {@link #epoch}; and call `@param cursorPositionEditor`(you can change `read position` and `mark deleted position` in this method).
+ * return `-1` if the result of compare and set `isResetting` is `false`, else return {@link #epoch}.
+ */
+public synchronized int endReset(Runnable cursorPositionEditor);
+
+/** Return false if there is a running epoch increase task or epoch was changed. **/
+public boolean isEpochOutdated(int expectedSpoch);
+
+public void getEpoch();
+```
+
+**CursorStats.java**
+```java
+public int epoch;
+public int isEpochIncreaseRunning;
+```
+
+A demo of `rewind` in Key_shared mode:
+```java
+if(cursor.epoch.startReset() != -1) {
```

Comment:
Your example is just an ideal situation since you call `dispatcher.clearDelayedMessageTracker(); `, these instructions need to be executed in the lock block of the Dispatcher, and you call `cursor.updatePosition()`, these instructions need to be executed in the lock block of the Cursor, It's not just Dispatcher who needs this feature, but Replication and others too. This will make the lock control rough. There are also asynchronous operations when dealing with these things, then the situation is more complicated. The current design is very flexible,  which is an advantage